aplib.net-demo/Assets/Scripts/Player/Movement.cs

```diff
@@ -0,0 +1,188 @@
+        Debug.DrawRay(transform.position, directionOnWall * 5, Color.blue);
+    }
+
+    /// <summary>
+    ///  Limit the player's velocity to the max speed.
+    ///  Also accounts for vertical velocity when on a slope.
+    /// </summary>
+    /// <param name="isOnWalkableSlope">Is the player on a walkable slope</param>
+    private void LimitVelocity(bool isOnWalkableSlope){
+        if (isOnWalkableSlope) {
+            if (_rigidbody.velocity.magnitude > _maxSpeed) _rigidbody.velocity = _rigidbody.velocity.normalized * _maxSpeed;   
+        } else {
+            Vector3 rigidbodyHorizontalVelocity = new(_rigidbody.velocity.x, 0, _rigidbody.velocity.z);
+            if (rigidbodyHorizontalVelocity.magnitude > _maxSpeed) {
+                Vector3 limitedVelocity = rigidbodyHorizontalVelocity.normalized * _maxSpeed;
+                _rigidbody.velocity = new Vector3(limitedVelocity.x, _rigidbody.velocity.y, limitedVelocity.z);
+            }
+        }
+        
+        // Draw a ray to visualize the player's velocity and direction
+        Debug.DrawRay(transform.position, _rigidbody.velocity, Color.red);
+    }
+
+    /// <summary>
+    /// Check if the player is on a walkable slope.
+    /// If so, return movement direction adjusted to the slope, and the hit point.
+    /// </summary>
+    /// <param name="directionOnSlope">Movement direction projected on the slope as a plane</param>
+    /// <param name="downHit">RaycastHit of the downward slope check</param>
+    /// <returns>A boolean that signifies whether the player is on a walkable slope or not</returns>
+    private bool OnWalkableSlope(out Vector3 directionOnSlope, out RaycastHit downHit) {
+        // Shoot a ray down to check if the player is on a slope
+        if (Physics.Raycast(transform.position, Vector3.down, out downHit, _playerHeight * 0.5f + 0.3f, _groundMask)) {
+            float angle = Vector3.Angle(Vector3.up, downHit.normal);
+            if (angle < _slopeAngle && angle != 0) {
+                directionOnSlope = Vector3.ProjectOnPlane(_horizontalVelocity, downHit.normal).normalized;
+                Debug.DrawRay(transform.position, directionOnSlope * 10, Color.blue);
+                return true;
+            }
+        }
+        
+        directionOnSlope = Vector3.zero;
+        return false;
+    }
+
+    /// <summary>
+    /// Check if the player is walking against a wall.
+    /// If so, return the wall's normal.
+    /// </summary>
+    /// <param name="wallNormal">The normal of the wall the player is walking against</param>
+    /// <returns>A boolean that signifies whether the player is walking against a wall or not</returns>
+    private bool WalkingAgainstWall(out Vector3 wallNormal) {
+        if (Physics.SphereCast(transform.position, _playerRadius, _horizontalVelocity, out RaycastHit hit, 0.5f, _groundMask)) {
+            wallNormal = hit.normal;
+            Debug.DrawRay(transform.position, wallNormal * 5, Color.green);
```

Comment:
Should this colour be defined somewhere so we have a uniform set of debug colour types?

(view rays are red, physics rays are blue etc etc that is an example but could be defined somewhere)